[Mobile] Unselect blocks using the hardware back button (Android)

geriux opened the pull request:
**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6483
- https://github.com/wordpress-mobile/WordPress-Android/pull/19828

## What?
This PR adds a new functionality on Android to allow unselecting blocks by using the hardware back button of the device.

## Why?
To prevent users from accidentally closing the editor when trying to unselect a block.

## How?
A few native changes were needed to make this implementation work on the host Android app:

A new event listener [was added](https://github.com/WordPress/gutenberg/pull/57279/commits/ca3588fbe7ba4d8461c0e1310d0bc1ec0f308aa5) to be able to communicate to the main app when it should trigger the default back button behavior instead of the implementations added within the React Native app. This listener is triggered within `invokeDefaultOnBackPressed` which will bubble up the event to the `EditPostActivity.java` file.

I used this [React Native guide](https://reactnative.dev/docs/integration-with-existing-apps#the-magic-reactrootview) as a reference on how to implement the back button code.

On the React side, I added a [BackHandler listener](https://github.com/WordPress/gutenberg/pull/57279/commits/acc52beb6eae94dddf208ddabc894b0c1acc4fd6) that will intercept when the back button is pressed and it will decide if it should take over the event or if it should skip it and trigger `invokeDefaultOnBackPressed`.

## Testing Instructions

> [!NOTE]  
> Use the [build from this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/19828#issuecomment-1865069203) to go through the testing cases.

### Unselecting a block

- Open the Android app
- Create a new post
- Add a Spacer block
- **Expect** to see the block selected
- Tap on the hardware back button
- **Expect** to see the Spacer block get unselected
- Tap on the hardware back button again
- **Expect** the editor to close

### Closing a modal

- Open the Android app
- Create a new post
- Tap on the "Hide keyboard button"
- Tap on the Inserter button
- Once the modal opens, tap on the hardware back button
- **Expect** to see the modal getting closed
- Tap on the hardware back button again
- **Expect** the editor to close

### Selecting and unselecting inner blocks

- Open the Android app
- Create a new post
- Add a Gallery block with a few images
- Select an image
- Tap on the hardware back button
- **Expect** to see no blocks selected
- Tap on an image again
- Then tap on the "Up arrow" to select the parent block
- Tap on the hardware back button
- **Expect** to see the block get unselected
- Tap on the hardware back button again
- **Expect** the editor to close 

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast <!-- if applicable -->

Unselecting a block|Closing a modal|Selecting and unselecting inner blocks|
-|-|-
<video src="https://github.com/WordPress/gutenberg/assets/4885740/f1814dfa-3454-4b71-94a2-4b0ebb362b81" width="250" />|<video src="https://github.com/WordPress/gutenberg/assets/4885740/e887a804-db7f-4263-a179-48a083c78b7f" width="250" />|<video src="https://github.com/WordPress/gutenberg/assets/4885740/0d967095-56ce-4f10-9969-7b538752b148" width="250" />

Commit:
Mobile - WPAndroidGlueCode - Implement BackHandler functionality

Commit:
Mobile - Editor Provider - Add BackHandler listener to unselect blocks

Commit:
Mobile - Jest setup - Adds BackHandler mocks

Commit:
Mobile - Edit post tests - Adds test for unselecting blocks using the hardware back button

fluiddot reviewed (commented):
I've gone through the testing instructions and worked as expected. However, I noticed the following minor issues. The first would be the only one that we might consider fixing before merging. The other two could be addressed (if needed) separately.

### Back button doesn't navigate back from Post settings screen

Seems an issue related to this PR as I can't reproduce it in version 23.9-rc-2.

https://github.com/WordPress/gutenberg/assets/14905380/320a8768-0324-4578-b7d5-9294a339c9a3

### Keyboard opens after navigating back from Help & Support

This is unrelated to this PR as I've been able to reproduce in `trunk`. My gut feeling is that it might be a regression introduced in https://github.com/WordPress/gutenberg/pull/57069. I'll open a GitHub issue as a follow-up.

https://github.com/WordPress/gutenberg/assets/14905380/cda98996-50e2-43a7-8aa0-a5edc2c93b4d

### Selecting and unselecting inner blocks

The current behavior looks good to me, but I wanted to share that my first impression was that navigating back would select the parent instead of unselecting the block. Absolutely not a blocker but I wonder what behavior would users expect 🤔.

Commit:
Track when it should handle the back press button for cases when there's another activity instanced and the editor is in the background

Commit:
Merge branch 'trunk' into feat/unselect-block-with-back-button

Commit:
Mobile - Keyboard Handling on Android - Set currentFocusedElement to null when hiding the soft keyboard. It also adds a check to not attempt to show the soft keyboard if there's no element set in currentFocusedElement

Commit:
Update Changelog

geriux commented:
> I've gone through the testing instructions and worked as expected. However, I noticed the following minor issues. The first would be the only one that we might consider fixing before merging. The other two could be addressed (if needed) separately.

Thank you for testing the build!

> Back button doesn't navigate back from Post settings screen

Nice catch! I've updated the code to track when it should intercept the back button and when it shouldn't, for this case, when the activity of the editor isn't visible it shouldn't intercept it. Addressed in https://github.com/WordPress/gutenberg/pull/57279/commits/ce40897a0654e9bb9c2c17c5581f60cd6ba4a525 and https://github.com/wordpress-mobile/WordPress-Android/pull/19828/commits/0960294747dde7ade8f9e485e41075250f23d713

https://github.com/WordPress/gutenberg/assets/4885740/f27f266f-f390-4e88-9fc0-b4a73b34dc14

> Keyboard opens after navigating back from Help & Support

Sorry for introducing this bug 😓 I have a potential fix in https://github.com/WordPress/gutenberg/pull/57279/commits/194c1f7a9c47d6256d2be51155b0e0357af53afa let me know what you think.

https://github.com/WordPress/gutenberg/assets/4885740/66f98437-0914-4578-8db9-3d3723b276e3

> Selecting and unselecting inner blocks

I'm hesitant about this, in cases where there are a few nesting levels, I worry that users might think they are stuck and unable to exit the editor after several presses of the back button. 😅 What do you think about shipping it as is and seeing if we receive any feedback about it?

Thanks again for testing!

fluiddot commented:
> I'm hesitant about this, in cases where there are a few nesting levels, I worry that users might think they are stuck and unable to exit the editor after several presses of the back button. 😅 What do you think about shipping it as is and seeing if we receive any feedback about it?

@geriux Sure, my comment was merely an observation and shouldn't block at all the PR. I'll review the PR and once approved, we could merge it as is.

fluiddot commented on packages/react-native-aztec/src/AztecInputState.js:
```diff
@@ -187,7 +187,10 @@ export const blurOnUnmount = ( element ) => {
-const dismissKeyboardDebounce = debounce( () => hideAndroidSoftKeyboard(), 0 );
+const dismissKeyboardDebounce = debounce( () => {
+	hideAndroidSoftKeyboard();
+	currentFocusedElement = null;
```
This makes me think that we might need to replicate in `blurOnUnmount` the behavior of `blur` to set the input in the same state. I.e. perform the following calls:
```
setCurrentCaretData( null );
notifyInputChange();
```

fluiddot replied:
I've debugged this further and noticed that the input state is not updated when an Aztec view is unmounted, i.e. the current focused element points to the removed reference. I'm working on a solution in a separate PR, but since it will follow the same approach implemented here we can leave this change as is.

fluiddot replied:
Here is the PR: https://github.com/WordPress/gutenberg/pull/57486. Let me know if you could take a look, thanks 🙇 !

fluiddot replied:
@geriux I've just merged https://github.com/WordPress/gutenberg/pull/57486. Once all CI checks pass, we can merge this PR. Thanks 🙇 !

geriux replied:
Awesome! 🚀 Thanks!

fluiddot commented on packages/react-native-bridge/index.js:
```diff
@@ -489,7 +494,12 @@ export function showAndroidSoftKeyboard() {
 
-	RNReactNativeGutenbergBridge.showAndroidSoftKeyboard();
+	const hasFocusedTextInput =
+		RCTAztecView.InputState.getCurrentFocusedElement() !== null;
```
Nitpick: [`isFocused` function](https://github.com/WordPress/gutenberg/blob/cc77161ebfc89667fd66848a670ca6fceebbee3b/packages/react-native-aztec/src/AztecInputState.js#L95-L97) already provides this value, hence we could simply call:
```suggestion
	const hasFocusedTextInput = RCTAztecView.InputState.isFocused();
```

geriux replied:
Updated in https://github.com/WordPress/gutenberg/pull/57279/commits/00597901147cff714288e3016233fa35097e6ece thanks for the suggestion!

fluiddot commented on packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java:
```diff
@@ -770,25 +783,28 @@ public void onPause(Activity activity) {
     }
 
     public void onDetach(Activity activity) {
+        mShouldHandleBackPress = false;
         mReactInstanceManager.onHostDestroy(activity);
         mRnReactNativeGutenbergBridgePackage.getRNReactNativeGutenbergBridgeModule().notifyModalClosed();
     }
 
     public void onDestroy(Activity activity) {
+        mShouldHandleBackPress = false;
```
NOTE: I debugged the app and seems `onPause` is called before detaching and destroying the activity, hence `mShouldHandleBackPress` would be always `false` at this point and we won't need to set it to `false` again. However, in case there might be other scenarios where `onPause` wouldn't be invoked, let's keep it as is.

fluiddot reviewed (approved):
LGTM 🎊 !

I checked the issues mentioned in https://github.com/WordPress/gutenberg/pull/57279#pullrequestreview-1794341923 and confirmed that it's working as expected.

- Back button doesn't navigate back from Post settings screen ✅ 
- Keyboard opens after navigating back from Help & Support ✅

Commit:
Use InputState isFocused method instead of manually checking if there's a focused element

Commit:
Merge branch 'trunk' into feat/unselect-block-with-back-button

# Conflicts:
#	packages/react-native-editor/CHANGELOG.md

fluiddot committed:
[RNMobile] Ensure Aztec input state function `blurOnUnmount` updates its state (#57486)

* Ensure `blurOnUnmount` updates internal input state

* Add unit test to cover `blurOnUnmount` logic
